src/utils/src/ExponentialBackoffRetryStrategy.c

```diff
@@ -0,0 +1,251 @@
+#include "Include_i.h"
+
+STATUS initializeDefaultExponentialBackoffConfig(PExponentialBackoffRetryStrategyConfig pExponentialBackoffRetryStrategyConfig) {
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+
+    CHK(pExponentialBackoffRetryStrategyConfig != NULL, STATUS_NULL_ARG);
+
+    pExponentialBackoffRetryStrategyConfig->maxRetryCount = KVS_INFINITE_EXPONENTIAL_RETRIES;
+    pExponentialBackoffRetryStrategyConfig->maxRetryWaitTime = HUNDREDS_OF_NANOS_IN_A_MILLISECOND * DEFAULT_KVS_MAX_WAIT_TIME_MILLISECONDS;
+    pExponentialBackoffRetryStrategyConfig->retryFactorTime = HUNDREDS_OF_NANOS_IN_A_MILLISECOND * DEFAULT_KVS_RETRY_TIME_FACTOR_MILLISECONDS;
+    pExponentialBackoffRetryStrategyConfig->minTimeToResetRetryState = HUNDREDS_OF_NANOS_IN_A_MILLISECOND * DEFAULT_KVS_MIN_TIME_TO_RESET_RETRY_STATE_MILLISECONDS;
+    pExponentialBackoffRetryStrategyConfig->jitterFactor = DEFAULT_KVS_JITTER_FACTOR_MILLISECONDS;
+
+CleanUp:
+    LEAVES();
+    return retStatus;
+}
+
+STATUS resetExponentialBackoffRetryState(PExponentialBackoffRetryStrategyState pExponentialBackoffRetryStrategyState) {
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+
+    CHK(pExponentialBackoffRetryStrategyState != NULL, STATUS_NULL_ARG);
+
+    pExponentialBackoffRetryStrategyState->currentRetryCount = 0;
+    pExponentialBackoffRetryStrategyState->lastRetryWaitTime = 0;
+    pExponentialBackoffRetryStrategyState->lastRetrySystemTime = 0;
+    pExponentialBackoffRetryStrategyState->status = BACKOFF_NOT_STARTED;
+    DLOGD("Resetting Exponential Backoff State");
+
+CleanUp:
+    LEAVES();
+    return retStatus;
+}
+
+STATUS exponentialBackoffRetryStrategyWithDefaultConfigCreate(PRetryStrategy* ppRetryStrategy) {
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    PExponentialBackoffRetryStrategyState pExponentialBackoffRetryStrategyState = NULL;
+
+    CHK(ppRetryStrategy != NULL, STATUS_NULL_ARG);
+
+    pExponentialBackoffRetryStrategyState = (PExponentialBackoffRetryStrategyState) MEMCALLOC(1, SIZEOF(ExponentialBackoffRetryStrategyState));
+    CHK(pExponentialBackoffRetryStrategyState != NULL, STATUS_NOT_ENOUGH_MEMORY);
+
+    CHK_STATUS(initializeDefaultExponentialBackoffConfig(&(pExponentialBackoffRetryStrategyState->exponentialBackoffRetryStrategyConfig)));
+    CHK_STATUS(resetExponentialBackoffRetryState(pExponentialBackoffRetryStrategyState));
+    DLOGD("Created exponential backoff retry strategy state with default configuration.");
+
+CleanUp:
+    if (STATUS_SUCCEEDED(retStatus)) {
+        (*ppRetryStrategy) = (PRetryStrategy)pExponentialBackoffRetryStrategyState;
+    } else {
+        SAFE_MEMFREE(pExponentialBackoffRetryStrategyState);
+    }
+
+    LEAVES();
+    return retStatus;
+}
+
+STATUS inRange(UINT64 value, UINT64 low, UINT64 high) {
```

Comment:
I am afraid the naming of this internal function might conflict with other libraries.

Reply:
Moreover, I would recommend implementing this as a macro (with the name in capitals)

Reply:
Updated to `checkConfigParameterRange`. I thought of macro but isn't the compiler smart enough to expand this inline? Just trying to avoid a macro.

Reply:
Many compilers will not be "smart" enough to inline it. This is not a biggie just macro is more preferred. Be careful with the macro as it uses the params which need to be wrapped in parenthesis and the invocation of the macro should use simple variables and not expressions (which can get re-evaluated)

Reply:
Okay I can change it to a macro (although I'm not super excited about macros :))